### PR TITLE
[MNG-6633] - Reduce memory usage of excludes

### PR DESCRIPTION
**Pull request by oehme**
ExcludesArtifactFilter was highly inefficient.
It took the group and artifact ID of an Exclusion, concatenated them into
a new String, which was kept in memory for the whole duration
of the build and then compared that String against the concatenation
of group and artifact ID of each incoming artifact, adding more
CPU cycles than necessary.

Instead we now just wrap the existing Exclusion object and check its
groupId and artifactId against the artifact to tell whether it should
be excluded.

The old class is kept around for binary compatibility, but is now unused.
Following this checklist to help us incorporate your 
contribution quickly and easily:

 - [x] Make sure there is a [JIRA issue](https://issues.apache.org/jira/browse/MNG) filed 
       for the change (usually before you start working on it).  Trivial changes like typos do not 
       require a JIRA issue.  Your pull request should address just this issue, without 
       pulling in other changes.
 - [x] Each commit in the pull request should have a meaningful subject line and body.
 - [x] Format the pull request title like `[MNG-XXX] - Fixes bug in ApproximateQuantiles`,
       where you replace `MNG-XXX` with the appropriate JIRA issue. Best practice
       is to use the JIRA issue title in the pull request title and in the first line of the 
       commit message.
 - [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [x] Run `mvn clean verify` to make sure basic checks pass. A more thorough check will 
       be performed on your pull request automatically.
 - [x] You have run the [Core IT][core-its] successfully.

If your pull request is about ~20 lines of code you don't need to sign an
[Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf) if you are unsure
please ask on the developers list.

To make clear that you license your contribution under 
the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)
you have to acknowledge this by using the following check-box.

 - [x] I hereby declare this contribution to be licenced under the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)

 - [ ] In any other case, please file an [Apache Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

[core-its]: https://maven.apache.org/core-its/core-it-suite/

### CODE REVIEW

**Comment by oehme** [comment]
In the 2000 module build I'm currently profiling, this shaves off a whole GB of heap usage.

**Comment by oehme** [comment]
Alternatively I could just create a new filter class and leave the old ones alone. However based on the existence of the getPatterns method, it seems some plugins actually depend on the concrete filter class used :/

**Comment by oehme** [comment]
Alternatively I could just create a new filter class and leave the old ones alone. However based on the existence of the getPatterns method, it seems some plugins actually depend on the concrete filter class used :/

**Comment by oehme** [comment]
Any idea if the concrete class and the getPatterns method matter? If not I could write a cleaner solution.

File: maven-core/src/main/java/org/apache/maven/artifact/resolver/filter/ExcludesArtifactFilter.java [hunk 1]
```diff
@@ -19,24 +19,43 @@
...
 
 /**
  * Filter to exclude from a list of artifact patterns.
  *
  * @author <a href="mailto:brett@apache.org">Brett Porter</a>
- * TODO I think this is equiv. to exclusion set filter in maven-core
  */
 public class ExcludesArtifactFilter
-    extends IncludesArtifactFilter
+        extends IncludesArtifactFilter
```
Comment (michael-o) [inline maven-core/src/main/java/org/apache/maven/artifact/resolver/filter/ExcludesArtifactFilter.java]: Superfluous indentation

File: maven-core/src/main/java/org/apache/maven/artifact/resolver/filter/IncludesArtifactFilter.java [hunk 2]
```diff
@@ -33,43 +33,43 @@
  * @author <a href="mailto:brett@apache.org">Brett Porter</a>
  */
 public class IncludesArtifactFilter
-    implements ArtifactFilter
+        implements ArtifactFilter
```
Comment (michael-o) [inline maven-core/src/main/java/org/apache/maven/artifact/resolver/filter/IncludesArtifactFilter.java]: same here

**Comment by oehme** [comment]
> As far as I understand the code, you just moved duplicate code into methods and classes, 

No, I didn't. The change is explained in detail in the PR message. Did you give that a read?

**Comment by michael-o** [comment]
Just read it, thanks. I am just surprised that a new class with new object consumes less memory than strings. I find sith the change.

**Comment by oehme** [comment]
The object has a header of 12 bytes, plus 4 bytes for both fields, rounded up to multiples of 8 => 24 bytes per object.

The string `commons-lang:commons-lang` has 25 characters, all LATIN-1, which on Java 9 and above is represented with 25 bytes (50 bytes on older Java versions). To that we need to add the array object header (12 bytes). We also have a String object wrapped around it, which has a reference to the char array (4 bytes) plus the `coder` field (4 bytes), plus the `hashCode` field (4 bytes) plus the object header (12 bytes) ==> 61 bytes for that String. It gets worse the longer a groupId/artifactId is. 

However, an even better solution would be to have a wrapper around the existing `Exclude` object. This is why I asked whether using a new class instead of the existing `ExcludeArtifactFilter` is an option (maybe @hboutemy can comment on that?). If we do that, we reduce the size per exclude to 16 bytes.

**Comment by michael-o** [comment]
That was the auto correction on Android ;-)

**Comment by oehme** [comment]
I thought you were taken by the dark side of the force :)

**Comment by oehme** [comment]
I've pushed a different approach to this problem which doesn't change the existing class at all and instead uses a new class. This is cleaner and uses even less memory. My only concern is that someone is actually doing `instanceof` checks on the filter in some plugin. Not sure if that's a realistic scenario :) If you agree, I'd prefer going with that solution.

**Comment by oehme** [comment]
I'll take that as a yes and removed the older commit :)

**Comment by rfscholte** [comment]
I just noticed you're actually touching maven-compat too. Is that required? I hope we can leave that artifact as it is, so there are more reasons to switch to Maven3 compatible code.

**Comment by oehme** [comment]
It's not required, but is it a problem? I was just changing it for consistency.

**Comment by rfscholte** [comment]
maven-compat contains classes that should not be used anymore in Maven 3. These classes are only there so Maven2 plugins still work. Hence we don't want to maintain these classes and they are likely to be removed in the next major version. 
Even though I understand the reason of consistency, by only doing improvements on the maintained code we're trying to push Maven2 plugin maintainers to migrate to Maven3 code.

**Comment by oehme** [comment]
I'm fine either way, I'll remove it.

**Comment by oehme** [comment]
Changes to maven-compat reverted.

**Comment by michael-o** [comment]
Can we finally merge it, we have agreed to accept this change. @slachiewicz you have assinged the ticket to yourself. Do you want to continue?

**Comment by slachiewicz** [comment]
Thanks for the reminder. I will make a merge in the next hour.

**Comment by slachiewicz** [comment]
I would also need to fix checkstyle issues (license header). 

Propably we should also do the same with this class: https://github.com/apache/maven/blob/ff8c5c8706497599e4fcf2f3ca3a768687991099/maven-core/src/main/java/org/apache/maven/artifact/resolver/filter/IncludesArtifactFilter.java#L35-L60

**Comment by michael-o** [comment]
@slachiewicz If this class has not been changed in this PR, then don't touch it. Focus on that modified parts only.

**Comment by jira-importer** [comment]
Resolve #8587

**Pull request closed without merging**